### PR TITLE
[2][resource reporting] Encapsulate poller and broadcaster into syncer in gcs

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This PR move the poller and broadcaster from gcs server to ray syncer. 

TODO in next PR: deprecate the code path of placement group resource reporting and move the broadcaster out of gcs cluster resource manager.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/ray_sync.h:36]: I like the API! It is also pretty easy to understand things (it is a bit sad poller has its own loop instead of directly calling poller->Poll() here, but it seems to be difficult). 

Is there any reason why we don't run poller in a separate thread too btw? I feel like this should give clear benefit in the large scale cluster, and seems to be low hanging fruit

Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:36]: poller is running in a separate thread right now. Btw, this is just the first step. We'll change it eventually to a push-based model like this prototype (https://github.com/ray-project/ray/pull/22073).

Comment (rkooo567) [reply src/ray/gcs/gcs_server/ray_sync.h:35]: Hmm you mean we will go back to push based report from pull based?

Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:35]: Right now, I think that's still arguable. So the plan is 1) make it lightweight (we need it anyway); 2) benchmark both to make the decision.  The pros are that with streaming, it seems way easier to do it in a push way.  Anyway, that might be very late.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:405]: @rkooo567  Btw, right now poller is not put in ray sync because of resource update. I'll put it there once I delete the legacy code.

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (rkooo567) [reply src/ray/gcs/gcs_server/ray_sync.h:35]: I see. This sounds reasonable. For other parts we can definitely discuss it again once it is ready to experiment

**Comment by iycheng** [comment]
working on fixing the test failures.

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (scv119) [inline src/ray/gcs/gcs_server/ray_sync.h:26]: the interface looks very clean.

nit: can you separate the implementation into -internal.h and add some comments on the interface?

Comment (jjyao) [inline src/ray/gcs/gcs_server/ray_sync.h:26]: nit: is it better to name it ResourceSyncer?

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (jjyao) [inline src/ray/gcs/gcs_server/ray_sync.h:36]: Move the impl to .cc file?

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (jjyao) [inline src/ray/gcs/gcs_server/ray_sync.h:46]: can we remove static unless we know there is a significant performance penalty to read config each time?

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (jjyao) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: In the comment, could you mention the thread model?

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (jjyao) [inline src/ray/gcs/gcs_server/ray_sync.h:110]: Who is using these? I didn't find the usage.

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:36]: Hmm, I think it's ok? Right now we don't have a standard for this I think. Btw, this will be updated in the follow up PR.

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:110]: Good catch!

File: src/ray/gcs/gcs_server/ray_sync.h
Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: It's going to be a general component. I don't want to bind it to Resource only. Later scheduling info is also going to use this component.
Do you have other suggestions for this?

Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: @scv119 The interface is still WIP since we need to refactor the code first and migrate it to the final version. It's going to be something similar like (https://github.com/ray-project/ray/pull/22073). Are you OK that I can update it in the later PR (it should be the PR [5])

- [3] Deprecate the old pg reporting code path
- [4] Remove syncer from gcs cluster resource manager and encapsulate everything in syncer module.
- [5] Introduce Reporter/Receiver API in the prototype and adaptor rayler and gcs with that.
- [6] Versioning the report
- [7] Experiment with protocol & communication layer change.

Comment (jjyao) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: In the long term, what information will be synced through this class? (I think we should at least rename sync to syncer)

Comment (jjyao) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: Do we have a design doc about the end state once all the PRs are merged? (for example, I'm curious about the reporter/receiver api and versioning)

Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: @jjyao this is only refactoring. Prototype is here is you'd like to have a check https://github.com/ray-project/ray/pull/22073

Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: It should be anything you think is worth broadcasting or sending.

Comment (jjyao) [reply src/ray/gcs/gcs_server/ray_sync.h:26]: oh, got it. I thought it was distributed versioning.

File: .bazelrc
Comment (rkooo567) [inline .bazelrc:11]: Is it temporary?

File: src/ray/gcs/gcs_server/gcs_resource_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_resource_manager.h:31]: Do we need this?

File: src/ray/gcs/gcs_server/gcs_resource_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_resource_manager.h:199]: Can you update docstring to explain what the syncer is doing?

File: src/ray/gcs/gcs_server/gcs_resource_report_poller.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_resource_report_poller.h:14]: oopss.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_server.cc:416]: what's the difference between dispatch vs post?

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_server.cc:426]: Do we need separate Initialize API? Why can't this be done within a constructor?

File: src/ray/gcs/gcs_server/grpc_based_resource_broadcaster.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/grpc_based_resource_broadcaster.cc:73]: Do we really need a broadcaster btw? It seems to not do much thing now

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/ray_syncer.h:112]: Good idea! I don't like on_update needs to post to the main thread now...

File: .bazelrc
Comment (iycheng) [reply .bazelrc:11]: Thanks for catching this.

File: src/ray/gcs/gcs_server/gcs_resource_manager.h
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_resource_manager.h:31]: Unfortunately, there is a circular dependence for now. It should be removed in the follow-up PR once we remove the pg resource reporting code path

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:416]: if it's already running in the event loop, it'll just execute this instead of posting to the queue.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:426]: Because of pg code path, we can't put them into syncer constructor. It'll be removed in later PR.

File: src/ray/gcs/gcs_server/grpc_based_resource_broadcaster.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/grpc_based_resource_broadcaster.cc:73]: I'm not changing the code for this PR. Just trying to move everything around with minimal change. Later I'll merge everything into syncer (see the prototype).

**Comment by iycheng** [comment]
@jjyao @scv119 @rkooo567  any further comment?

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (scv119) [inline src/ray/gcs/gcs_server/ray_syncer.h:47]: we need some comments to describe what this function does. it's non trivial at the moment.

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (scv119) [inline src/ray/gcs/gcs_server/ray_syncer.h:78]: should update `NodeResourceChange` and `ResourcesData` belong to separate method? looks they are doing different things.

i think one nice thing of documenting the APIs: if we can not easily describe the APIs by simple comments, probably we should separate them

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (scv119) [inline src/ray/gcs/gcs_server/ray_syncer.h:97]: similar for AddNode/RemoveNode

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (scv119) [inline src/ray/gcs/gcs_server/ray_syncer.h:28]: need a bit more detail to help understand what's it doing now.

File: src/ray/gcs/gcs_server/ray_syncer.h
Comment (iycheng) [reply src/ray/gcs/gcs_server/ray_syncer.h:78]: @scv119  this is not the final one we'll have. I'll comment it there.

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_resource_manager.cc:27]: @scv119  this will be removed in 4th PR. I make it a raw pointer so that I don't need to update unit tests for this one.

The specific component of poller/broadcaster is tested, but not integrated with this manager right now.

I'll leave it as it is for now. Once we created the good form of RaySyncer [within 2weeks].

I'll add test cases for that. For now, it's ok I think since it's not making things worse and we have nightly and CI to cover the errors.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:418]: std::move a const&?

**Review by jjyao — APPROVED** [review]
lgtm for now. I'll review it again after the follow-up PRs to clean up the pg report path.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:418]: @jjyao  it's about moving report from the capture in the lambda.

**Review by scv119 — APPROVED** [review]
Chatted with Yi offline. This is good to go assuming this is temporary.

**Comment by iycheng** [comment]
test failure is not related. merge this one.

**Pull request merged**
